pilot/pkg/xds/debuggen.go

```diff
@@ -0,0 +1,279 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/golang/protobuf/ptypes/any"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	status "github.com/envoyproxy/go-control-plane/envoy/service/status/v3"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	v3 "istio.io/istio/pilot/pkg/xds/v3"
+	"istio.io/pkg/log"
+)
+
+const (
+	// TypeDebug requests debug info from istio, a secured implementation for istio debug interface
+	TypeDebug = "istio.io/debug"
+)
+
+var activeSystemDebuggers = []string{
+	"registryz",
+	"endpointz",
+	"endpointShardz",
+	"cachez",
+	"configz",
+	"sidecarz",
+	"resourcesz",
+	"instancesz",
+	"authorizationz",
+	"push_status",
+	"pushcontext",
+	"connections",
+	"inject",
+	"mesh",
+	"syncz",
+	"adsz",
+}
+
+var activeNamespaceDebuggers = []string{
+	"config_dump",
+	"ndsz",
+	"syncz",
+	"edsz",
+	"config_distribution",
+}
+
+// DebugGen is a Generator for istio debug info
+type DebugGen struct {
+	Server *DiscoveryServer
+}
+
+func NewDebugGen(s *DiscoveryServer) *DebugGen {
+	return &DebugGen{
+		Server: s,
+	}
+}
+
+// Generate XDS debug responses according to the incoming debug request
+func (dg *DebugGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates *model.PushRequest) (model.Resources, error) {
+	res := []*any.Any{}
+	var err error
+	var out []byte
+	if w.ResourceNames == nil {
+		return res, fmt.Errorf("debug type is required")
+	}
+	resourceName := w.ResourceNames[0]
```

Comment:
if we only allow exactly 1 we should probably give an error if there are multiple?

Reply:
we will allow more than 1 resourceName in case we need to pass the proxyID, eg `istioctl x debug helloworld-v1.xxxx.sample`  in such case, we will use the sa from this specific pod and also will use it as proxyID.

Reply:
we can use query params

Reply:
how can we pass query params in the discoveryRequest?

Reply:
Its just a string, it can be anything. Like `edsz?proxyID=foo`

Reply:
I see, you mean combine then into one single string then we can have only one ResourceNames